### ISSUE TITLE
byres selection fails with insertion code

### ISSUE THREAD

**Issue by RMeli** [MDAnalysis/mdanalysis#2672]
## Expected behavior ##

Correct `byres` selection for PDB files.

## Actual behavior ##

Using `byres` with PDB files containing insertion codes leads to erroneous selections, where all residues with the same number but different insertion codes are selected.

This is linked to #2308.

## Code to reproduce the behavior ##

``` python
In [1]: import MDAnalysis as mda

In [2]: u = mda.Universe("p.pdb")

In [3]: ui = mda.Universe("p_icode.pdb")

In [4]: u.select_atoms("byres(around 2.0 resnum 13)")
Out[4]: <AtomGroup with 21 atoms>

In [5]: ui.select_atoms("byres(around 2.0 resnum 13)")
Out[5]: <AtomGroup with 12 atoms>
```

with `p_icode.pd`

```
ATOM    217  N   GLU L  13      -5.997  11.588  22.006  1.00 22.34           N
ATOM    218  H   GLU L  13      -7.033  11.673  22.032  1.00  0.00           H
ATOM    219  CA  GLU L  13      -5.303  10.472  22.709  1.00 24.14           C
ATOM    220  HA  GLU L  13      -4.258  10.727  22.888  1.00  0.00           H
ATOM    221  C   GLU L  13      -5.380   9.199  21.913  1.00 23.81           C
ATOM    222  O   GLU L  13      -6.470   8.998  21.302  1.00 25.39           O
ATOM    223  CB  GLU L  13      -6.150  10.222  23.967  1.00 26.06           C
ATOM    224  HB1 GLU L  13      -7.200  10.170  23.680  1.00  0.00           H
ATOM    225  HB2 GLU L  13      -5.849   9.274  24.412  1.00  0.00           H
ATOM    226  CG  GLU L  13      -5.978  11.320  24.994  1.00 36.69           C
ATOM    227  HG1 GLU L  13      -4.919  11.576  25.032  1.00  0.00           H
ATOM    228  HG2 GLU L  13      -6.548  12.185  24.654  1.00  0.00           H
ATOM    229  CD  GLU L  13      -6.421  11.003  26.391  1.00 42.27           C
ATOM    230  OE1 GLU L  13      -6.995   9.892  26.528  1.00 46.63           O
ATOM    231  OE2 GLU L  13      -6.208  11.780  27.305  1.00 46.31           O
ATOM    232  N   ASP L  14      -4.407   8.330  21.972  1.00 22.45           N
ATOM    233  H   ASP L  14      -3.528   8.586  22.465  1.00  0.00           H
ATOM    234  CA  ASP L  14      -4.526   6.972  21.347  1.00 21.73           C
ATOM    235  HA  ASP L  14      -5.031   6.955  20.381  1.00  0.00           H
ATOM    236  C   ASP L  14      -5.408   6.153  22.306  1.00 20.49           C
ATOM    237  O   ASP L  14      -5.818   6.528  23.416  1.00 19.90           O
ATOM    238  CB  ASP L  14      -3.184   6.429  20.919  1.00 20.07           C
ATOM    239  HB1 ASP L  14      -3.360   5.540  20.314  1.00  0.00           H
ATOM    240  HB2 ASP L  14      -2.694   7.189  20.311  1.00  0.00           H
ATOM    241  CG  ASP L  14      -2.229   6.045  22.052  1.00 17.98           C
ATOM    242  OD1 ASP L  14      -2.698   5.227  22.871  1.00 21.20           O
ATOM    243  OD2 ASP L  14      -1.046   6.434  21.950  1.00 20.58           O
ATOM    244  N   LYS L  14A     -5.864   5.031  21.805  1.00 21.06           N
ATOM    245  H   LYS L  14A     -5.496   4.733  20.879  1.00  0.00           H
ATOM    246  CA  LYS L  14A     -6.839   4.178  22.443  1.00 21.93           C
ATOM    247  HA  LYS L  14A     -7.640   4.846  22.760  1.00  0.00           H
ATOM    248  C   LYS L  14A     -6.366   3.509  23.720  1.00 23.19           C
ATOM    249  O   LYS L  14A     -7.278   3.215  24.547  1.00 24.43           O
ATOM    250  CB  LYS L  14A     -7.402   3.138  21.440  1.00 26.01           C
ATOM    251  HB1 LYS L  14A     -7.879   3.658  20.609  1.00  0.00           H
ATOM    252  HB2 LYS L  14A     -6.587   2.520  21.063  1.00  0.00           H
```
and `p.pdb`

```
ATOM    217  N   GLU L  13      -5.997  11.588  22.006  1.00 22.34           N
ATOM    218  H   GLU L  13      -7.033  11.673  22.032  1.00  0.00           H
ATOM    219  CA  GLU L  13      -5.303  10.472  22.709  1.00 24.14           C
ATOM    220  HA  GLU L  13      -4.258  10.727  22.888  1.00  0.00           H
ATOM    221  C   GLU L  13      -5.380   9.199  21.913  1.00 23.81           C
ATOM    222  O   GLU L  13      -6.470   8.998  21.302  1.00 25.39           O
ATOM    223  CB  GLU L  13      -6.150  10.222  23.967  1.00 26.06           C
ATOM    224  HB1 GLU L  13      -7.200  10.170  23.680  1.00  0.00           H
ATOM    225  HB2 GLU L  13      -5.849   9.274  24.412  1.00  0.00           H
ATOM    226  CG  GLU L  13      -5.978  11.320  24.994  1.00 36.69           C
ATOM    227  HG1 GLU L  13      -4.919  11.576  25.032  1.00  0.00           H
ATOM    228  HG2 GLU L  13      -6.548  12.185  24.654  1.00  0.00           H
ATOM    229  CD  GLU L  13      -6.421  11.003  26.391  1.00 42.27           C
ATOM    230  OE1 GLU L  13      -6.995   9.892  26.528  1.00 46.63           O
ATOM    231  OE2 GLU L  13      -6.208  11.780  27.305  1.00 46.31           O
ATOM    232  N   ASP L  14      -4.407   8.330  21.972  1.00 22.45           N
ATOM    233  H   ASP L  14      -3.528   8.586  22.465  1.00  0.00           H
ATOM    234  CA  ASP L  14      -4.526   6.972  21.347  1.00 21.73           C
ATOM    235  HA  ASP L  14      -5.031   6.955  20.381  1.00  0.00           H
ATOM    236  C   ASP L  14      -5.408   6.153  22.306  1.00 20.49           C
ATOM    237  O   ASP L  14      -5.818   6.528  23.416  1.00 19.90           O
ATOM    238  CB  ASP L  14      -3.184   6.429  20.919  1.00 20.07           C
ATOM    239  HB1 ASP L  14      -3.360   5.540  20.314  1.00  0.00           H
ATOM    240  HB2 ASP L  14      -2.694   7.189  20.311  1.00  0.00           H
ATOM    241  CG  ASP L  14      -2.229   6.045  22.052  1.00 17.98           C
ATOM    242  OD1 ASP L  14      -2.698   5.227  22.871  1.00 21.20           O
ATOM    243  OD2 ASP L  14      -1.046   6.434  21.950  1.00 20.58           O
ATOM    244  N   LYS L  15     -5.864   5.031  21.805  1.00 21.06           N
ATOM    245  H   LYS L  15     -5.496   4.733  20.879  1.00  0.00           H
ATOM    246  CA  LYS L  15     -6.839   4.178  22.443  1.00 21.93           C
ATOM    247  HA  LYS L  15     -7.640   4.846  22.760  1.00  0.00           H
ATOM    248  C   LYS L  15     -6.366   3.509  23.720  1.00 23.19           C
ATOM    249  O   LYS L  15     -7.278   3.215  24.547  1.00 24.43           O
ATOM    250  CB  LYS L  15     -7.402   3.138  21.440  1.00 26.01           C
ATOM    251  HB1 LYS L  15     -7.879   3.658  20.609  1.00  0.00           H
ATOM    252  HB2 LYS L  15     -6.587   2.520  21.063  1.00  0.00           H
```

## Current version of MDAnalysis ##

- Which version are you using? 0.20.1
- Which version of Python (`python -V`)? 3.7.6
- Which operating system? macOS

**Comment by IAlibay** [comment]
Please correct me if I'm wrong, I _think_ this is related to #2669?

**Comment by RMeli** [comment]
Yes, I completely missed #2669. Different problem but same underlying cause.

**Comment by IAlibay** [comment]
I find this particularly case rather entertaining to overthink though. As far as I am aware `byres` probably is doing it's job, since ASP14 and LYS14A are technically both residues 14... 

As Richard alluded to (https://github.com/MDAnalysis/mdanalysis/issues/2669#issuecomment-618267827), an implementation of `bryes` that uses `residinces` instead (`byresidx`? `byridx`? `residx`?) would be advantageous here.

**Comment by lilyminium** [comment]
The documentation says `byres` is synonymous to `same residue as`, so I think this counts as an error. As I pointed out in #2669 we already have `same resid as` for the resid bits. Either way as long as we have *one* method that works, that'll do.

https://github.com/MDAnalysis/mdanalysis/blob/400f93cd94626ae94de0fa5ab283beac5aa3ad40/package/MDAnalysis/core/selection.py#L1013-L1029

As far as I can tell it should be as easy as changing this to `resindices` and `segindices`? If we want to overengineer it and allow for custom attributes, the TopologyAttr metaclass could populate the dictionary on class instantiation instead.

I don't really understand what's going on here though. When you select from the universe with the icodes you get *fewer* atoms? Shouldn't it be flipped, i.e. "byres" picks up both Asp14 and Lys14A in the `ui` universe, but only Asp14 in the `u` universe?

**Comment by RMeli** [comment]
> I don't really understand what's going on here though. When you select from the universe with the icodes you get fewer atoms?

Absolutely! I had the two file names inverted when I run the code snippet and forgot to fix it. This is the correct output, sorry for the confusion:

```python
In [1]: import MDAnalysis as mda

In [2]: u = mda.Universe("p.pdb")

In [3]: ui = mda.Universe("p_icode.pdb")

In [4]: u.select_atoms("byres(around 2.0 resnum 13)")
Out[4]: <AtomGroup with 12 atoms>

In [5]: ui.select_atoms("byres(around 2.0 resnum 13)")
Out[5]: <AtomGroup with 21 atoms>
```

**Comment by RMeli** [comment]
```python
In [1]: import MDAnalysis as mda

In [2]: u = mda.Universe("p.pdb")

In [3]: ui = mda.Universe("p_icode.pdb")

In [4]: u.select_atoms("same residue as (around 2.0 resnum 13)")
Out[4]: <AtomGroup with 12 atoms>

In [5]: ui.select_atoms("same residue as (around 2.0 resnum 13)")
Out[5]: <AtomGroup with 21 atoms>
```

Changing the mapping to `'residue': 'resindices'` works as expected:

```python
In [1]: import MDAnalysis as mda

In [2]: u = mda.Universe("p.pdb")

In [3]: ui = mda.Universe("p_icode.pdb")

In [4]: u.select_atoms("same residue as (around 2.0 resnum 13)")
Out[4]: <AtomGroup with 12 atoms>

In [5]: ui.select_atoms("same residue as (around 2.0 resnum 13)")
Out[5]: <AtomGroup with 12 atoms>
```

**Comment by RMeli** [comment]
I can confirm that the following mapping
```python
prop_trans = {
        'residue': 'resindices',
        'segment': 'segindices',
        'resname': 'resnames',
        'resid': 'resindices',
        'segid': 'segindices',
    }
```
and changing `byres` to use `resindices` as follows
```python
class ByResSelection(UnarySelection):
    token = 'byres'
    precedence = 1

    def apply(self, group):
        res = self.sel.apply(group)
        unique_res = unique_int_1d(res.resindices.astype(np.int64))
        mask = np.in1d(group.resindices, unique_res)

        return group[mask].unique
```
solve the problems outlined here and in #2669.
```python
In [1]: import MDAnalysis as mda

In [2]: u = mda.Universe('a.psf', 'a.dcd')

In [3]: ag = u.select_atoms('byres resname TIP3')

In [4]: print(set(ag.resnames))
{'TIP3'}
```

Mapping both `'residue': 'resindices',` and `'resid': 'resindices',` is redundant (as noted in #2669), but I think that leaving `resid` selection will lead to wrong results in edge cases (see #2308). The changes do not seem to break any code since `resindices` are used internally by `byres` or `same` .

@lilyminium unless you want to work on this (since you reported the problem first and suggested the solution) I can open a PR (where we can discuss this further).

**Comment by lilyminium** [comment]
@RMeli feel free to open a PR but I don't think mapping `resid` to `resindices` and especially `segid` to `segindices` is a good idea. The main issue with #2308 seems to be the mismatch between the selection "resid 30A" and the `resid` attribute, which can be fixed by adding a new attribute that combines the residue numbers and the insertion codes. Changing the type and contents of `resid` would be very disruptive. I'm biased in this discussion because I primarily use `resids` over `resnums` but I doubt I'm alone, especially since all of analysis modules use `resids` over `resnums`.

**Comment by RMeli** [comment]
@lilyminium I complete agree that changing the content of `resid` is not feasible nor desirable at this stage (although I still think that an ID should be an unique identifier; for the residue number only we already have `resnum`).

My suggestion of `resid` to `resindices` mapping was motivated by the fact that the intended use of the selection `"same resid as (around 2.0 resnum 13)"` is to select residues near `resnum 13`; including all the residues with the same `resid` (i.e. `resnum` since they are the same) and different `icode` could sometimes end up selecting residues very far away from `resnum 13` (I'm thinking of systems with 10-15 insertion codes). However, `same resid as` is not only used in combination with `around` therefore I think leaving it as it is, as you suggest, is probably the safest option?

**Comment by lilyminium** [comment]
@RMeli that sounds good! It would be nice to have the `"same resid as"` selection work the way it looks in the very rare case that someone does need to select all residues with the same `resid` across icodes and chains and wrapped-around PDBs and what-have-you, especially since `"same residue as"` and `"byres"` will both exist to select the actual residue.

**Comment by RMeli** [comment]
> It would be nice to have the "same resid as" selection work the way it looks in the very rare case that someone does need to select all residues with the same resid across icodes and chains and wrapped-around PDBs and what-have-you.

Sure! But I think this could also be achieved by `same resnum as`, or am I missing something?

**Comment by lilyminium** [comment]
I generally believe that things should do what they look like they do. In this case, I think `"same resid as ..."` should return all atoms with the same `.resid` attribute as `...`, and in fact that should happen for all topology attributes (which could be unambiguously achieved by using `TopologyAttrMeta` to update `SameSelection.prop_trans`). If every topology attribute *except* `resid` returned the matching values in the actual attribute, that means the API is inconsistent. Again, I don't use `resnum` because I know that `resid` is just an alias of it, so if `same resid as` wasn't giving me what I expected, my first thought would be "the code is broken" rather than "I should try `same resnum as`".

Basically I expect selection strings and masking to give me the same results:

```python
ag = u.select_atoms('around 3.0 protein')
sel = u.select_atoms('same resid as (around 3.0 protein)')
mask = u.residues[np.isin(u.residues.resids, ag.resids)]
sel == mask
```

and I think any deviation from that should have really good reasons.

**Comment by RMeli** [comment]
Your point on masking is a really good one and I completely agree that your code snipped should work. And I also agree that changing `resid` to `resindices` is dangerous since there would be a mismatch between the selection `same resid as` and what it is actually returned.

I guess the point I'm trying to make, also in #2308, is that `resid` should not be a simple alias of `resnum`, but should be an unique identifier of the residue (in many cases they will match since a residue can often be uniquely identified by its residue number; but in some cases this is not true). It would be useful to have a _unique_ residue identifier, easier to use than `resindex`.

**Comment by lilyminium** [comment]
Yes, you're totally right that the implementation of the `resid` label doesn't match up with the name and it should be different to `resnum`. It has just been around for so long that I feel the easiest solution is to make a new attribute and go with that instead, possibly deprecating `.resid` (and think very hard about the current uses of `.resid` in the code, the dihedral selections should probably not use them). It does currently fit in with the numeric `atom.id` and how VMD uses `resid`.
